DAOS-12107 control: Verbose output for daos container list

0xE0F opened the pull request:
Adds verbose output as follow:

```
$:~/src/daos/install/bin> sudo ./daos container list -V denisb
UUID                                 Label   Type    ObjClass RF
----                                 -----   ----    -------- --
c29a1242-c77a-486d-904f-1a6b321e2e0a nvme    POSIX   SX       0
33dc4dd4-259c-411b-a93b-ba83e18d642c nfs     POSIX   EC_4P2G2 0
ee224070-5589-4aa9-9847-5bd2dc8a99e9 scratch unknown          0
```

Signed-off-by: Denis Barakhtanov <denis.barahtanov@croit.io>

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficent testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

0xE0F committed:
DAOS-12107 control: verbose output for daos container list

Adds verbose output as follow:

$:~/src/daos/install/bin> sudo ./daos container list -V denisb
UUID                                 Label   Type    ObjClass RF
----                                 -----   ----    -------- --
c29a1242-c77a-486d-904f-1a6b321e2e0a nvme    POSIX   SX       0
33dc4dd4-259c-411b-a93b-ba83e18d642c nfs     POSIX   EC_4P2G2 0
ee224070-5589-4aa9-9847-5bd2dc8a99e9 scratch unknown          0

Signed-off-by: Denis Barakhtanov <denis.barahtanov@croit.io>

mjmac commented:
Created #10905 to test from an internal branch.

mjmac commented:
> Created #10905 to test from an internal branch.

Passed all tests. Pylint failure seems to be some kind of issue on master.

mjmac commented:
@0xE0F: PR looks good, but I think you'll need to rebase it on the latest master in order to bring in some changes to the container query output. I'll push the updated PR to an internal branch for full testing.

0xE0F committed:
Rebase and a bit more concise version

Signed-off-by: Denis Barakhtanov <denis.barahtanov@croit.io>

0xE0F commented:
@mjmac Thank you !
I rebased to the latest master and updated to a bit more concise version.

Maybe it's a bit of a stretch, but would the team consider using verbose output as a default one ?
Or the additional RPCs to get info are the deal breaker ?

Sample of output:
```
denis@dev:~/daos/install/bin> sudo ./daos container list -V denisb
UUID                                 Label    Type    ObjClass DirClass FileClass RF
----                                 -----    ----    -------- -------- --------- --
1daf60cc-1725-4694-9abc-f4f3b4da3c06 scratch2 POSIX   UNKNOWN  S1       S1        0
fb7c9ae6-b088-45d3-94f7-7e48a7746d6f nfs      POSIX   UNKNOWN  SX       SX        0
1c668e65-ab19-498f-83f3-35a879ce09b1 nvme     POSIX   UNKNOWN  SX       SX        0
59dc1615-aa01-43a3-b479-07fbce117fe6 scratch  unknown                             0
094e32de-7739-4789-bc19-bb518070e636 scratch3 POSIX   UNKNOWN  SX       S1        0
```

mchaarawi reviewed (changes requested):
I have some concerns with this PR.
some of the information printed in verbose mode is specific to only POSIX containers. I don't think this should be printed with the list containers command. 
I'm talking about all the object classes which only apply to POSIX container, are not container properties but more like just data from the generic container list perspective.
so my suggestion is for the container list in verbose mode to print only container properties (RF, Rlevel, etc.) but not include type specific data for containers like the object class for posix containers.
This also appears to cause overhead by mounting the container again from the go code on daos cont create since query is run from there. this can all be avoided if you exclude that type specific data.

mchaarawi commented on src/control/cmd/daos/container.go:
```diff
@@ -562,14 +589,33 @@ func (cmd *containerListCmd) Execute(_ []string) error {
+	contInfo := make([]*containerInfo, 0, len(contIDs))
+	for _, contID := range contIDs {
+		if cmd.Verbose {
+			info, err := cmd.containerInfo(&contID.UUID)
```
In a production setting where a pool will have containers owned by different users, isn't this going to fail everytime with ENOPERM for containers they don't have access to?
so i question the usefulness in this case.

0xE0F replied:
fair enough, I'm not using multiple owners at the moment.
On the other hand, the ones without access can be omitted, or perhaps reading container attributes allowed for anyone ?

mchaarawi replied:
> On the other hand, the ones without access can be omitted

But that is not really a container list anymore (and not even a verbose list), is it? it would very weird that list-containers return one thing, and list-containers --verbose returns a subset. sounds like you need a completely different option to return containers that you are an owner (or have read permissions on) and return those.  

> perhaps reading container attributes allowed for anyone

the problem is that attributes you are reading here for POSIX containers are not really attributes from the container perspective, but just regular data (POSIX containers store them in the SB object which is just a regular object from container permissions sake) so it involves a regular object fetch that you should have read permissions for.

mjmac replied:
@kccain pointed out yesterday that there is a new-ish container API that might be useful here: https://github.com/daos-stack/daos/blob/11e2bddb4575df47092a91ec9cc23973a0b5d067/src/include/daos_pool.h#L611

In theory, that should allow us to construct a filter to only show containers meeting certain criteria. I'm not sure if there is a way to express read permission as a filter criterion, though.

kccain replied:
Not related to container owners -  but applicable to outputting some common properties: This new interface combines list containers behavior with the basic part of a container query (the daos_cont_info_t information).  However it does not return properties. I tried to leave a placeholder in one of the data structures in case we might decide to try to add properties in later, but didn't think too exactly about how it would be implemented. Except to think that perhaps a more sophisticated implementation of the cart bulk transfer could be done -  i.e., list_cont_bulk_create() that in this implementation currently refers to a single linear buffer of size ncont * sizeof(struct daos_pool_cont_info2). Maybe the pci_reserved[] pointers in that structure could have some additional dynamically allocated storage to accommodate properties coming back from the engine in the bulk transfer.

0xE0F replied:
Looks like permission issue might get on the way with the verbose output.
I do agree with @mchaarawi that skipping containers without proper access does not express `list` command semantic at all.

I guess tooling have to deal with containers one by one 😭 
Maybe one day with another interface it would be straightforward to implement 😬

0xE0F commented:
> 

yeah, I do agree that printing attributes that are not part of the container type not the best idea.
The whole point of this PR is to make a container management a bit more convenient to use in case when there are dozens of container with different types and options.
And it's a bit hard to find that specific container with posix type, rf=1 and oclass=S1 when all you have in the output is label and UUID.
This leads to give a container labels like `smb_rf1_S1`, `scratch_rf0_SX` etc. 
Or list attributes for each container until one is found.
Same goes for the shelling out and use `daos` cmd in the pipeline with other tools.

What about grouping container by type? Maybe something like this:

```
denis@dev:~/daos/install/bin> sudo ./daos container list -V denisb
POSIX containers
----------------

UUID                                 Label    ObjClass DirClass FileClass RF
----                                 -----    -------- -------- --------- --
1daf60cc-1725-4694-9abc-f4f3b4da3c06 scratch2 UNKNOWN  S1       S1        0
fb7c9ae6-b088-45d3-94f7-7e48a7746d6f nfs      UNKNOWN  SX       SX        0
1c668e65-ab19-498f-83f3-35a879ce09b1 nvme     UNKNOWN  SX       SX        0
094e32de-7739-4789-bc19-bb518070e636 scratch3 UNKNOWN  SX       S1        0


OTHER containers
----------------

UUID                                 Label    RF
----                                 -----    --
59dc1615-aa01-43a3-b479-07fbce117fe6 scratch  1
```
